### DISCUSSION TITLE
Creating GADP client stub

### DISCUSSION THREAD

**Discussion by behunin** [NationalSecurityAgency/ghidra#4845] [Debugger]
I am looking into creating a GADP client stub for the yuzu emulator and just want to see if this is a good idea and if so then what is the correct way to start the connection. Usually I would gather this information from docs but I can't find any. 

[As you can see here](https://github.com/yuzu-emu/yuzu/tree/master/src/core/debugger)  there is already a GDB stub for the project. I know Ghidra works with GDB but I am just curious as to what I can do with GADP.

**Comment by d-millar** [comment]
@behunin I’m not sure what the use care here would be that would merit the work, particularly given the existence of the gdbstub.  The primary use cases for GADP are two: (1) it isolates the JVM running Ghidra proper from the JVM running the debugger agent, so that a crash in the debugger will not destroy any current session/work in Ghidra, and (2) it facilitates a server/client setup when you need the agent to run on a different machine.  While you could write a GADP server to support your target, you would essentially have no model to build on, as the existing agent models assume the existence of some underlying native debugger.

**Reply by behunin** [reply]
> the existing agent models assume the existence of some underlying native debugger.

The gdbstub in that project is just the communication layer for its native debugger implementation, so I figured that would be sufficient for GADP to do the same thing with ghidra.

Your telling me that is not its design purpose, but is it possible to use GADP this way?

**Reply by d-millar** [reply]
Entirely possible - just not something we would recommend.

**Reply by d-millar** [reply]
Entirely possible - just not something we would recommend.

**Reply by behunin** [reply]
I fail to see why you do not recommend this other than, that's not how you designed it.
So I still wish to pursue this endeavor and I would like some guidance please.

**Reply by d-millar** [reply]
@behunin Well, you asked if it was a good idea....  That said, assuming you still wanted to pursue this, you will need to implement the basic protocol, the server-side handler for it, and some flavor of DebuggerObjectModel.  Given your target, you would presumably be re-implementing something similar to the gdb variant but in C rather than Java.  This would essentially mean porting the code in Ghidra/Debug/Debugger-agent-gdb and Ghidra/Debug/Debugger-gadp to C. Obviously, you could pare this down a bit to the functionality you wanted, but it's still a large fraction of this code.

**Reply by behunin** [reply]
Thank you @d-millar 
I wanted to make sure I vetted this idea just to make sure it wasn't a futile pursuit.

Implementing basic protocol is trivial with boost asio and C++ files generated with protoc.
I will port GadpClient to C++ and see where that gets me.

**Reply by d-millar** [reply]
Just to be clear, you need to implement GadpServer and a model.  The protocol will be trivial; the model will not.
